### PR TITLE
Overhaul scoped identifier rules and add dependent name support

### PR DESCRIPTION
**Pull request by jdrouhard**
!! Breaking change due to restructuring and renaming of scoped identifiers !!

Scoped identifier rules have been renamed to "qualified_x_identifier" and all variants are aliased so they appear as simply "qualified_identifier" in the syntax tree. This name better reflects the c++ standard's language describing this concept.

Qualified identifier rules are now always the "parent" with a right associative rule that lets them nest indefinitely. Each rule terminates with some kind of non-scope identifier.

This commit also adds support for the 'template' keyword where required to disambiguate dependent template names. Field accessors (., ->) can be followed by keyword template before a template method. The scope resolution operator can be immediately followed by the 'template' keyword to indicate the following dependent name with '<' is actually a template and shouldn't be parsed as a binary operator. These keywords are required for dependent names in these contexts.

https://en.cppreference.com/w/cpp/language/identifiers#Qualified_identifiers
https://en.cppreference.com/w/cpp/language/dependent_name

### CODE REVIEW

**Comment by jdrouhard** [comment]
@maxbrunsfeld This significantly improves scoped identifier parsing, but does introduce a breaking change for queries. Let me know what you think. This is one of the last major things that I've noticed is broken on modern c++ codebases and is one of the remaining big ticket items I wanted to get to.

@theHamsta - Take a look at the changes in the various test files (and the highlights query file in this change) to get a sense of what might need to be updated in nvim-treesitter.

Basically, `template_type`, `template_function` and `template_method` no longer have nested scoped types. The scoped identifier rule is always the parent with a template type being either the "namespace" (lhs) or "name" (rhs). Nested scopes will have more scoped type identifier rules being the "namespace" (rhs).

**Comment by jdrouhard** [comment]
@SmiteshP you may also be interested in this for nvim-gps. This change would require a tweak to your query.

**Comment by SmiteshP** [comment]
> @SmiteshP you may also be interested in this for nvim-gps. This change would require a tweak to your query.

Thanks for the heads up! 😄

File: grammar.js [hunk 1]
```diff
@@ -858,66 +854,67 @@ module.exports = grammar(C, {
...
       )
     ),
 
-    scoped_field_identifier: $ => prec(1, seq(
+    dependent_identifier: $ => seq('template', $.template_function),
+    dependent_field_identifier: $ => seq('template', $.template_method),
+    dependent_type_identifier: $ => seq('template', $.template_type),
+
+    _scoped_namespace: $=> prec(1, seq(
       field('namespace', optional(choice(
```
Comment (jdrouhard) [inline grammar.js:867]: Since this is a breaking change anyway, I'm leaning toward renaming this field 'scope' since they are not always namespaces. The parser can't actually know the semantic type, but it may not necessarily be a namespace. It could be a type, template type, dependent name, etc.

File: queries/highlights.scm [hunk 2]
```diff
@@ -11,8 +11,7 @@
   name: (field_identifier) @function)
 
 (template_function
-  name: (scoped_identifier
-    name: (identifier) @function))
+  name: (identifier) @function)
```
Comment (jdrouhard) [inline queries/highlights.scm:14]: Queries will be easier to define now that template_function, template_method, and template_type no longer have nested scoped types for the actual type. It's simply the rightmost (non-scoped) identifier.

File: grammar.js [hunk 3]
```diff
@@ -858,66 +854,67 @@ module.exports = grammar(C, {
...
+    )),
+
+    scoped_field_identifier: $ => seq(
+      $._scoped_namespace,
       field('name', choice(
-        $._field_identifier,
-        $.operator_name,
-        $.destructor_name
+        alias($.dependent_field_identifier, $.dependent_name),
+        $.scoped_field_identifier,
```
Comment (jdrouhard) [inline grammar.js:879]: I am also considering aliasing the various "scoped_identifier" type rules so they are all simply "scoped_identifier" instead of having the 4 different rule names depending on the context and type of the inner-most rhs identifier. Thoughts?

File: grammar.js (hunk 1 above)
Comment (maxbrunsfeld) [reply grammar.js:867]: Makes sense.

File: grammar.js (hunk 3 above)
Comment (maxbrunsfeld) [reply grammar.js:879]: I think it’s a good idea.

**Review by maxbrunsfeld — APPROVED** [review]

**Comment by jdrouhard** [comment]
Last change since you reviewed @maxbrunsfeld - I also changed `scoped_identifier` to `qualified_identifier` (and aliased all references to the various `qualified_x_identifier` rules to simply refer to them as `qualified_identifier`). 

This better reflects exactly what the c++ standard refers to these things as and makes it easier to understand what the grammar/syntax tree is referring to when nodes of this type appear. I'd like to limit breaking changes to as few commits as possible, so doing this now is preferred to renaming later.

https://en.cppreference.com/w/cpp/language/identifiers#Qualified_identifiers

**Comment by maxbrunsfeld** [comment]
Yeah, that sounds good. I’m cool with making fairly sweeping changes, if it will make this parser more useful to people who are experienced with C++ and it’s tooling. It shouldn’t be too hard for folks to adjust their queries.
